Add Check Support for Java 21 Record Pattern Syntax : PatternVariableName

child of https://github.com/checkstyle/checkstyle/issues/14890
I have read check documentation:https://checkstyle.org/checks/naming/patternvariablename.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```

PS D:\CS\test> cat src/Test2.java                                                
record ColoredPoint(int p, int x, String c) { }
record Rectangle(ColoredPoint upperLeft, ColoredPoint lowerRight) { }
public class Test2 {
    void test(Object obj) {
        if (!(obj instanceof Rectangle(ColoredPoint POINT,ColoredPoint ppPOINT))) {  // violation
        }
    }

}
PS D:\CS\test> cat config.xml    
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
        <module name="PatternVariableName">

        </module>

    </module>
</module>
PS D:\CS\test> java  -jar checkstyle-10.17.0-all.jar -c config.xml src/Test2.java
Starting audit...
[ERROR] D:\CS\test\src\Test2.java:5:53: Name 'POINT' must match pattern '^[a-z][a-zA-Z0-9]*$'. [PatternVariableName]
Audit done.
Checkstyle ends with 1 errors.
PS D:\CS\test> 
```

Describe what you expect in detail.

input file

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/PatternVariableNameCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/PatternVariableNameCheckTest.java
@@ -120,4 +120,24 @@ public class PatternVariableNameCheckTest
                         "InputPatternVariableNameUnnamed.java"),
                 expected);
     }
+
+    @Test
+    public void testPatternVariableNameRecordPattern() throws Exception {
+
+        final String pattern = "^([a-z][a-zA-Z0-9]*|_)$";
+
+        final String[] expected = {
+            "15:36: " + getCheckMessage(MSG_INVALID_PATTERN, "XX", pattern),
+            "15:44: " + getCheckMessage(MSG_INVALID_PATTERN, "__", pattern),
+            "20:49: " + getCheckMessage(MSG_INVALID_PATTERN, "S", pattern),
+            "25:28: " + getCheckMessage(MSG_INVALID_PATTERN, "XX", pattern),
+            "25:36: " + getCheckMessage(MSG_INVALID_PATTERN, "__", pattern),
+            "31:41: " + getCheckMessage(MSG_INVALID_PATTERN, "S", pattern),
+        };
+
+        verifyWithInlineConfigParser(
+                getNonCompilablePath(
+                        "InputPatternVariableNameRecordPattern.java"),
+                expected);
+    }
 }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/naming/patternvariablename/InputPatternVariableNameRecordPattern.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/naming/patternvariablename/InputPatternVariableNameRecordPattern.java
@@ -1,0 +1,39 @@
+/*
+PatternVariableName
+format = (default)^([a-z][a-zA-Z0-9]*|_)$
+
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.naming.patternvariablename;
+
+public class InputPatternVariableNameRecordPattern {
+
+    void test(Object o) {
+        if (o instanceof Point(int x, int _)) {}
+        if (o instanceof Point(int XX, int __)) {}
+        // 2 violations above:
+        //                'Name 'XX' must match pattern*.'
+        //                'Name '__' must match pattern*.'
+        if (o instanceof ColoredPoint(Point(int x, int yy), String color)) {}
+        if (o instanceof ColoredPoint(_, String S)) {}
+        //violation above, 'Name 'S' must match pattern*.'
+    }
+    void test2(Object o) {
+        switch (o) {
+            case Point(int XX, int __) when XX > 0-> {}
+            // 2 violations above:
+            //                'Name 'XX' must match pattern*.'
+            //                'Name '__' must match pattern*.'
+            case Point(int x, int _) -> {}
+            case ColoredPoint(Point(int x, int yy), String color) -> {}
+            case ColoredPoint(_, String S) -> {}
+            // violation above, 'Name 'S' must match pattern*.'
+            default -> {}
+        }
+    }
+
+    record Point(int x, int y) { }
+    record ColoredPoint(Point point, String color) { }
+}

